### PR TITLE
Add lowering for fraction_max_pool2d

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120460

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @desertfire

### CODE REVIEW

**Comment by amjames** [comment]
I still need to add some tests, and run some benchmarks to establish the conditions where we should fallback.

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -4384,6 +4383,110 @@ def adaptive_max_pool2d(x, output_size):
...
+@register_lowering(aten.fractional_max_pool2d)
+def fractional_max_pool2d(x, kernel_size, output_size, random_samples):
+    x.realize_hint()
+    *batch, inp_h, inp_w = x.get_size()
+    inp_h = V.graph.sizevars.evaluate_static_shape(inp_h)
+    inp_w = V.graph.sizevars.evaluate_static_shape(inp_w)
+    kernel_h, kernel_w = kernel_size
+    h_out, w_out = output_size
+
+    breakpoint()
```
Comment (lezcano) [inline torch/_inductor/lowering.py:4426]: oops

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -4384,6 +4383,110 @@ def adaptive_max_pool2d(x, output_size):
...
+
+    return load
+
+
+@register_lowering(aten.fractional_max_pool2d)
+def fractional_max_pool2d(x, kernel_size, output_size, random_samples):
+    x.realize_hint()
+    *batch, inp_h, inp_w = x.get_size()
+    inp_h = V.graph.sizevars.evaluate_static_shape(inp_h)
+    inp_w = V.graph.sizevars.evaluate_static_shape(inp_w)
```
Comment (lezcano) [inline torch/_inductor/lowering.py:4422]: Do we need these?

**Review by lezcano — COMMENTED** [review]
How difficult would it be to merge this lowering and `avg_max_pool2d`?

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (amjames) [reply torch/_inductor/lowering.py:4422]: It looks like we do 1) The lowering for other pooling ops has this, and 2) testing it out it looks like that size can't be symbolic to use it as the `size` for an indirect indexing expression later.

Comment (lezcano) [reply torch/_inductor/lowering.py:4422]: For 1), the other pooling ops may very well be improved as well lol.
2) What do you mean by "indrirect indexing cannot be symbolic"? See
https://github.com/pytorch/pytorch/blob/80900e2172f8b7b18ef44750a92bea27b51dd606/torch/_inductor/lowering.py#L2870

File: test/inductor/test_torchinductor.py [hunk 3]
```diff
@@ -3216,6 +3216,63 @@ def fn(x):
         )
         assertGeneratedKernelCountEqual(self, 0)
 
+    @config.patch(fallback_random=True)
```
Comment (peterbell10) [inline test/inductor/test_torchinductor.py:3219]: Would it make more sense to pass the `_random_samples` argument to `fractional_max_pool2d_with_indices` so we remove the random generation from the equation?

File: torch/_meta_registrations.py [hunk 4]
```diff
@@ -4316,6 +4316,82 @@ def meta_max_pool2d_with_indices(
...
+@register_meta(aten.fractional_max_pool2d.default)
+def meta_fractional_max_pool2d(self_, kernel_size, output_size, random_samples):
+    torch._check(
+        self_.ndim in (3, 4),
+        lambda: f"fractional_max_pool2d: Expected 3D or 4D tensor, but got: {self_._ndim}",
+    )
+    ndim = self_.ndim
+
+    for i in range(3):
+        d = i + (ndim - 3)
```
Comment (peterbell10) [inline torch/_meta_registrations.py:4328]: NIT: 
```suggestion
    for d in range(ndim - 3, ndim):
```

File: torch/_meta_registrations.py [hunk 5]
```diff
@@ -4316,6 +4316,82 @@ def meta_max_pool2d_with_indices(
...
+        torch._check(
+            self_.size(d) > 0,
+            f"fractional_max_pool2d: Expected input to have non-zero "
+            f" size for non-batch dimenions, but got {self_.size()} with dimension {d} empty",
+        )
+
+    # the check and message are out of sync, but this matches the structured meta
```
Comment (peterbell10) [inline torch/_meta_registrations.py:4335]: This is fine, single ints should be broadcasted by the python bindings.

File: torch/_meta_registrations.py [hunk 6]
```diff
@@ -4316,6 +4316,82 @@ def meta_max_pool2d_with_indices(
...
+    inW = self_.size(-1)
+    if ndim == 4:
+        nBatch = self_.size(0)
+    else:
+        nBatch = 1
+
+    samples_shape = [nBatch, nPlane, 2]
+
+    torch._check(
+        all(ss == random_samples.size(i) for i, ss in enumerate(samples_shape)),
```
Comment (peterbell10) [inline torch/_meta_registrations.py:4358]: Slightly weird, but `fractional_max_pooling_check_shape` seems to allow the batch dimension to be `>=` on  `random_samples`. Also the error message for dimensional mismatch will be this `.size` call raising which isn't ideal. Eager has a an explicit dim check.

File: torch/_inductor/lowering.py [hunk 7]
```diff
@@ -4384,6 +4383,104 @@ def adaptive_max_pool2d(x, output_size):
...
+            samples_loader([*prefix, dim]), size=3, check=False
+        )
+
+        seq_i = ops.index_expr(
+            sympy.floor((i + sample) * alpha) - sympy.floor(sample * alpha), torch.int64
+        )
+
+        mask = ops.lt(
+            ops.index_expr(i, torch.int64),
+            ops.sub(ops.index_expr(out_sz, torch.int64), ops.constant(1, torch.int64)),
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4409]: NIT:
```suggestion
            ops.index_expr(out_sz - 1, torch.int64),
```

File: torch/_inductor/lowering.py [hunk 8]
```diff
@@ -4384,6 +4383,104 @@ def adaptive_max_pool2d(x, output_size):
...
+    samples_loader = samples.make_loader()
+
+    def load(prefix, i):
+        sample = ops.indirect_indexing(
+            samples_loader([*prefix, dim]), size=3, check=False
+        )
+
+        seq_i = ops.index_expr(
+            sympy.floor((i + sample) * alpha) - sympy.floor(sample * alpha), torch.int64
+        )
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4405]: This works, but it's a bit weird to do `indirect_indexing` followed by `index_expr`. We can just do everythign directly in the inductor IR and avoid `indirect_indexing`:
```suggestion
        sample = samples_loader([*prefix, dim])
        i_expr = ops.index_expr(i, samples.get_dtype())
        alpha_expr = ops.index_expr(alpha, samples.get_dtype())
        seq_i = ops.floor((i_expr + sample) * alpha) - ops.floor(sample * alpha)
        seq_i = ops.to_dtype(seq_i, torch.int64)
```

File: torch/_inductor/lowering.py [hunk 9]
```diff
@@ -4384,6 +4383,104 @@ def adaptive_max_pool2d(x, output_size):
...
+        for ih, iw in itertools.product(range(kernel_size[0]), range(kernel_size[1])):
+            val = x_loader([*prefix, h_start_index + ih, w_start_index + iw])
+            if return_index:
+                index = ops.index_expr(
+                    (h_start_index + ih) * inp_w + w_start_index + iw, torch.int64
+                )
+                if maxindex is None:
+                    maxindex = index
+                else:
+                    maxindex = ops.where(ops.gt(val, maxval), index, maxindex)
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:4457]: This needs logic to handle nans and ties. 
https://github.com/pytorch/pytorch/blob/03662bd4bc4da297a5f1352bc0c624a77e27d7e4/aten/src/ATen/native/cuda/FractionalMaxPool2d.cu#L79-L83

**Review by peterbell10 — DISMISSED** [review]

**Comment by peterbell10** [comment]
Looks good, just needs some benchmark results.

**Review by lezcano — APPROVED** [review]
Great, thanks!

**Comment by lezcano** [comment]
but yeah, let's have a look at the benchmarks first.

**Comment by amjames** [comment]
Benchmarks: 

```[----------------------------- fractional_max_pool2d-cuda -----------------------------]
                                                                    |  Compile  |  Eager
32 threads: ----------------------------------------------------------------------------
      (36, 3, 128, 128), fractional_max_pool2d((2, 2), (32, 32))    |      30   |     27
      (36, 3, 128, 128), fractional_max_pool2d((3, 3), (32, 32))    |      35   |     29
      (36, 3, 128, 128), fractional_max_pool2d((4, 4), (32, 32))    |      37   |     34
      (36, 3, 128, 128), fractional_max_pool2d((5, 5), (32, 32))    |      34   |     37
      (36, 3, 128, 128), fractional_max_pool2d((2, 2), (64, 64))    |      45   |     47
      (36, 3, 128, 128), fractional_max_pool2d((3, 3), (64, 64))    |      45   |     50
      (36, 3, 128, 128), fractional_max_pool2d((4, 4), (64, 64))    |      45   |     54
      (36, 3, 128, 128), fractional_max_pool2d((5, 5), (64, 64))    |      62   |     65
      (36, 3, 128, 512), fractional_max_pool2d((2, 2), (32, 128))   |      68   |     70
      (36, 3, 128, 512), fractional_max_pool2d((3, 3), (32, 128))   |      92   |     94
      (36, 3, 128, 512), fractional_max_pool2d((4, 4), (32, 128))   |     118   |    118
      (36, 3, 128, 512), fractional_max_pool2d((5, 5), (32, 128))   |     121   |    123
      (36, 3, 128, 512), fractional_max_pool2d((2, 2), (64, 256))   |     172   |    173
      (36, 3, 128, 512), fractional_max_pool2d((3, 3), (64, 256))   |     172   |    185
      (36, 3, 128, 512), fractional_max_pool2d((4, 4), (64, 256))   |     173   |    211
      (36, 3, 128, 512), fractional_max_pool2d((5, 5), (64, 256))   |     252   |    255
      (36, 3, 512, 128), fractional_max_pool2d((2, 2), (128, 32))   |      70   |     73
      (36, 3, 512, 128), fractional_max_pool2d((3, 3), (128, 32))   |      93   |     95
      (36, 3, 512, 128), fractional_max_pool2d((4, 4), (128, 32))   |     118   |    119
      (36, 3, 512, 128), fractional_max_pool2d((5, 5), (128, 32))   |     121   |    123
      (36, 3, 512, 128), fractional_max_pool2d((2, 2), (256, 64))   |     172   |    173
      (36, 3, 512, 128), fractional_max_pool2d((3, 3), (256, 64))   |     171   |    179
      (36, 3, 512, 128), fractional_max_pool2d((4, 4), (256, 64))   |     171   |    195
      (36, 3, 512, 128), fractional_max_pool2d((5, 5), (256, 64))   |     241   |    243
      (36, 3, 512, 512), fractional_max_pool2d((2, 2), (128, 128))  |     261   |    263
      (36, 3, 512, 512), fractional_max_pool2d((3, 3), (128, 128))  |     356   |    358
      (36, 3, 512, 512), fractional_max_pool2d((4, 4), (128, 128))  |     459   |    460
      (36, 3, 512, 512), fractional_max_pool2d((5, 5), (128, 128))  |     464   |    466
      (36, 3, 512, 512), fractional_max_pool2d((2, 2), (256, 256))  |     678   |    676
      (36, 3, 512, 512), fractional_max_pool2d((3, 3), (256, 256))  |     676   |    716
      (36, 3, 512, 512), fractional_max_pool2d((4, 4), (256, 256))  |     679   |    834
      (36, 3, 512, 512), fractional_max_pool2d((5, 5), (256, 256))  |    1010   |   1010

Times are in microseconds (us).

```

**Review by peterbell10 — APPROVED** [review]

**Comment by amjames** [comment]
I did forget to annotate, Kernel size (5, 5) or larger is expected to fallback to eager, otherwise I am checking for exactly one kernel generated. The benchmark checks for an expected number of kernels to be generated for each shape/args, otherwise it throws. 


```

[------------------------------ fractional_max_pool2d-cuda -----------------------------]
                                                                     |  Compile  |  Eager
32 threads: -----------------------------------------------------------------------------
      (36, 3, 128, 128), fractional_max_pool2d((2, 2), (32, 32))     |      40   |     27
      (36, 3, 128, 128), fractional_max_pool2d((3, 3), (32, 32))     |      35   |     29
      (36, 3, 128, 128), fractional_max_pool2d((4, 4), (32, 32))     |      36   |     34
      (36, 3, 128, 128), fractional_max_pool2d((5, 5), (32, 32))*    |      34   |     37
      (36, 3, 128, 128), fractional_max_pool2d((2, 2), (64, 64))     |      45   |     47
      (36, 3, 128, 128), fractional_max_pool2d((3, 3), (64, 64))     |      45   |     50
      (36, 3, 128, 128), fractional_max_pool2d((4, 4), (64, 64))     |      45   |     54
      (36, 3, 128, 128), fractional_max_pool2d((5, 5), (64, 64))*    |      63   |     65
      (36, 3, 128, 512), fractional_max_pool2d((2, 2), (32, 128))    |      68   |     70
      (36, 3, 128, 512), fractional_max_pool2d((3, 3), (32, 128))    |      92   |     94
      (36, 3, 128, 512), fractional_max_pool2d((4, 4), (32, 128))    |     118   |    119
      (36, 3, 128, 512), fractional_max_pool2d((5, 5), (32, 128))*   |     121   |    124
      (36, 3, 128, 512), fractional_max_pool2d((2, 2), (64, 256))    |     172   |    173
      (36, 3, 128, 512), fractional_max_pool2d((3, 3), (64, 256))    |     172   |    184
      (36, 3, 128, 512), fractional_max_pool2d((4, 4), (64, 256))    |     174   |    211
      (36, 3, 128, 512), fractional_max_pool2d((5, 5), (64, 256))*   |     254   |    256
      (36, 3, 512, 128), fractional_max_pool2d((2, 2), (128, 32))    |      70   |     73
      (36, 3, 512, 128), fractional_max_pool2d((3, 3), (128, 32))    |      93   |     95
      (36, 3, 512, 128), fractional_max_pool2d((4, 4), (128, 32))    |     118   |    119
      (36, 3, 512, 128), fractional_max_pool2d((5, 5), (128, 32))*   |     121   |    123
      (36, 3, 512, 128), fractional_max_pool2d((2, 2), (256, 64))    |     172   |    173
      (36, 3, 512, 128), fractional_max_pool2d((3, 3), (256, 64))    |     172   |    179
      (36, 3, 512, 128), fractional_max_pool2d((4, 4), (256, 64))    |     171   |    196
      (36, 3, 512, 128), fractional_max_pool2d((5, 5), (256, 64))*   |     241   |    243
      (36, 3, 512, 512), fractional_max_pool2d((2, 2), (128, 128))   |     261   |    263
      (36, 3, 512, 512), fractional_max_pool2d((3, 3), (128, 128))   |     356   |    358
      (36, 3, 512, 512), fractional_max_pool2d((4, 4), (128, 128))   |     459   |    460
      (36, 3, 512, 512), fractional_max_pool2d((5, 5), (128, 128))*  |     466   |    468
      (36, 3, 512, 512), fractional_max_pool2d((2, 2), (256, 256))   |     678   |    676
      (36, 3, 512, 512), fractional_max_pool2d((3, 3), (256, 256))   |     676   |    716
      (36, 3, 512, 512), fractional_max_pool2d((4, 4), (256, 256))   |     682   |    839
      (36, 3, 512, 512), fractional_max_pool2d((5, 5), (256, 256))*  |    1010   |   1010

Times are in microseconds (us).

*: Expected fallback to eager for large kernel size
```

**Comment by lezcano** [comment]
Ah, that makes sense, thanks!

**Pull request closed without merging**